common/src/com/thoughtworks/go/config/materials/git/GitMaterial.java

```diff
@@ -106,28 +127,32 @@ public MaterialConfig config() {
     }
 
     public List<Modification> modificationsSince(File baseDir, Revision revision, final SubprocessExecutionContext execCtx) {
+        unshallowIfNeeded(ProcessOutputStreamConsumer.inMemoryConsumer(), revision, baseDir);
         GitCommand gitCommand = getGit(baseDir);
         return gitCommand.modificationsSince(revision);
     }
 
     public MaterialInstance createMaterialInstance() {
-        return new GitMaterialInstance(url.forCommandline(), branch, submoduleFolder, UUID.randomUUID().toString());
+        return new GitMaterialInstance(url.forCommandline(), branch, submoduleFolder, UUID.randomUUID().toString(), shallowClone);
     }
 
     @Override
     protected void appendCriteria(Map<String, Object> parameters) {
         parameters.put(ScmMaterialConfig.URL, url.forCommandline());
         parameters.put("branch", branch);
+        parameters.put("shallowClone", shallowClone);
```

Comment:
I guess another option instead of migrating fingerprints is to write extra configuration validation to ensure consistency. "Validation Error: You can not have multiple git materials to same repository and branch with different shallowClone value". But this seem throwing more burdens to the user's end and limiting the use of the feature.